File: Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemWithBadge.m
/*
* Backpack - Skyscanner's Design System
*
* Copyright © 2020 Skyscanner Ltd. All rights reserved.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*   http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

#import "BPKHorizontalNavigationItemWithBadge.h"
#import "BPKHorizontalNavigationBadge.h"

#import <Backpack/BPKColor.h>
#import <Backpack/BPKLabel.h>
#import <Backpack/Spacing.h>

NS_ASSUME_NONNULL_BEGIN
@interface BPKHorizontalNavigationItemWithBadge ()
@property(nonatomic, strong) UIView *containerView;
@property(nonatomic, strong) BPKHorizontalNavigationBadge *badge;
@property(nonatomic, strong) BPKLabel *titleLabel;
@property(nonatomic, strong, readonly) UIColor *contentColor;
@end

@implementation BPKHorizontalNavigationItemWithBadge

- (instancetype)initWithTitle:(NSString *)title badgeMessage:(NSString *)badgeMessage {
    self = [super initWithFrame:CGRectZero];

    if (self) {
        self.badge.message = badgeMessage;
        self.titleLabel.text = title;
        [self setupSubViews];
    }

    return self;
}

- (void)setupSubViews {
    [self addSubview:self.containerView];
    [self.containerView addSubview:self.badge];
    [self.containerView addSubview:self.titleLabel];
    [self setContentCompressionResistancePriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisHorizontal];

    [NSLayoutConstraint activateConstraints:@[
        [self.titleLabel.leadingAnchor constraintEqualToAnchor:self.containerView.leadingAnchor constant:0],
        [self.titleLabel.topAnchor constraintEqualToAnchor:self.containerView.topAnchor],
        [self.titleLabel.bottomAnchor constraintEqualToAnchor:self.containerView.bottomAnchor],
        [self.badge.leadingAnchor constraintEqualToAnchor:self.titleLabel.trailingAnchor constant:BPKSpacingMd],
        [self.badge.topAnchor constraintEqualToAnchor:self.containerView.topAnchor],
        [self.badge.bottomAnchor constraintEqualToAnchor:self.containerView.bottomAnchor],
        [self.containerView.trailingAnchor constraintEqualToAnchor:self.badge.trailingAnchor constant:0],
        [self.containerView.centerYAnchor constraintEqualToAnchor:self.centerYAnchor],
        [self.containerView.centerXAnchor constraintEqualToAnchor:self.centerXAnchor],
        [self.containerView.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.leadingAnchor constant:BPKSpacingBase],
        [self.trailingAnchor constraintGreaterThanOrEqualToAnchor:self.containerView.trailingAnchor constant:BPKSpacingBase]
    ]];
}

- (void)updateStyle {
    BPKFontStyle fontStyle = BPKFontStyleTextBaseEmphasized;
    if (self.size == BPKHorizontalNavigationSizeSmall) {
        fontStyle = BPKFontStyleTextSmEmphasized;
    }

    self.titleLabel.fontStyle = fontStyle;
    self.titleLabel.textColor = self.contentColor;
    self.badge.color = self.contentColor;
}

#pragma mark - property getters/setters

- (void)setSelected:(BOOL)selected {
    [super setSelected:selected];

    [self updateStyle];
}

- (void)setSize:(BPKHorizontalNavigationSize)size {
    if (_size != size) {
        _size = size;

        [self updateStyle];
    }
}

- (void)setAppearance:(BPKHorizontalNavigationAppearance)appearance {
    if (_appearance != appearance) {
        _appearance = appearance;
        
        [self updateStyle];
    }
}

- (UIColor *)contentColor {
    if (self.selected) {
        if (self.selectedColor != nil) {
            return self.selectedColor;
        }
        return BPKColor.primaryColor;
    } else {
        switch (self.appearance) {
            case BPKHorizontalNavigationAppearanceNormal:
                return BPKColor.textPrimaryColor;
                break;
            case BPKHorizontalNavigationAppearanceAlternate:
                return BPKColor.skyGrayTint07;
        }
    }
}

- (BPKLabel *)titleLabel {
    if (_titleLabel == nil) {
        BPKFontStyle fontStyle = BPKFontStyleTextBaseEmphasized;
        if (self.size == BPKHorizontalNavigationSizeSmall) {
            fontStyle = BPKFontStyleTextSmEmphasized;
        }

        _titleLabel = [[BPKLabel alloc] initWithFontStyle:fontStyle];
        _titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    }

    return _titleLabel;
}

- (BPKHorizontalNavigationBadge *)badge {
    if (_badge == nil) {
        _badge = [[BPKHorizontalNavigationBadge alloc] initWithFrame:CGRectZero];
        _badge.userInteractionEnabled = NO;
        _badge.translatesAutoresizingMaskIntoConstraints = NO;
    }

    return _badge;
}

- (UIView *)containerView {
    if (_containerView == nil) {
        _containerView = [UIView new];
        _containerView.userInteractionEnabled = NO;
        _containerView.translatesAutoresizingMaskIntoConstraints = NO;
    }

    return _containerView;
}

@end
NS_ASSUME_NONNULL_END

File: Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemWithBadge.swift
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018-2021 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

class BPKHorizontalNavigationItemWithBadge<Size: BPKHorizontalNavigationSize>: UIControl, BPKHorizontalNavigationItem {

    override var isSelected: Bool {
        didSet {
            updateStyle()
        }
    }

    dynamic var selectedColor: UIColor? {
        didSet {
            updateStyle()
        }
    }

    var appearance: BPKHorizontalNavigationAppearance = .normal {
        didSet {
            updateStyle()
        }
    }

    private(set) var showNotificationDot: Bool = false

    private lazy var titleLabel: BPKLabel = {
        let label = BPKLabel(fontStyle: Size.fontStyle)
        label.translatesAutoresizingMaskIntoConstraints = false
        label.textColor = contentColor

        return label
    }()

    private lazy var badge: BPKHorizontalNavigationBadge = {
        let badge = BPKHorizontalNavigationBadge(frame: .zero)
        badge.translatesAutoresizingMaskIntoConstraints = false
        badge.isUserInteractionEnabled = false
        badge.color = contentColor

        return badge
    }()

    private lazy var containerView: UIStackView = {
        let view = UIStackView()
        view.translatesAutoresizingMaskIntoConstraints = false
        view.isUserInteractionEnabled = false
        view.spacing = BPKSpacingMd
        view.axis = .horizontal
        view.distribution = .fillProportionally
        view.alignment = .center

        return view
    }()

    private var contentColor: UIColor {
        if isSelected {
            return selectedColor ?? .bpk_primary
        } else {
            switch appearance {
            case .normal: return .bpk_textPrimary
            case .alternate: return .bpk_skyGrayTint07
            }
        }
    }

    init(title: String, badgeMessage: String) {
        super.init(frame: .zero)

        setUp(title: title, badgeMessage: badgeMessage)
    }

    private func notificationLocation(for labelFrame: CGRect) -> CGPoint {
        if BPKRTLSupport.viewIsRTL(self) {
            return CGPoint(x: labelFrame.minX - BPKSpacingMd, y: labelFrame.minY - BPKSpacingSm)
        } else {
            return CGPoint(x: labelFrame.maxX, y: labelFrame.minY - BPKSpacingSm)
        }
    }

    private func setUp(title: String, badgeMessage: String) {
        addSubview(containerView)
        containerView.addArrangedSubview(titleLabel)
        containerView.addArrangedSubview(badge)
        badge.message = badgeMessage
        titleLabel.text = title

        setContentCompressionResistancePriority(.required, for: .horizontal)

        NSLayoutConstraint.activate([
            containerView.leadingAnchor.constraint(greaterThanOrEqualTo: leadingAnchor, constant: BPKSpacingBase),
            containerView.topAnchor.constraint(equalTo: topAnchor),
            trailingAnchor.constraint(greaterThanOrEqualTo: containerView.trailingAnchor, constant: BPKSpacingBase),
            containerView.bottomAnchor.constraint(equalTo: bottomAnchor),
            containerView.centerXAnchor.constraint(equalTo: centerXAnchor)
        ])

        updateStyle()
    }

    private func updateStyle() {
        titleLabel.textColor = contentColor
        badge.color = contentColor
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}
